server/lib/scheduler/Schedule.ts

```diff
@@ -0,0 +1,142 @@
+import { map } from "asyncro";
+import * as _ from "lodash";
+import { UserSchedule } from "./UserSchedule";
+
+class Schedule {
+  // the following object maps root workspace ids to timestamps
+  // I didn't use a Map (and map a workspace object to a timestamp)
+  // because Sequelize doesn't preserve object identity across queries
+  private cacheForWhenTreeLastWorkedOn = {};
+
+  private rootParentCache;
+  private schedule = new Map;
+  private timeLimit;
+
+  public constructor({ rootParentCache, timeLimit }) {
+    this.rootParentCache = rootParentCache;
+    this.timeLimit = timeLimit;
+  }
+
+  private doesUserHaveASchedule(userId) {
+    return this.schedule.has(userId);
+  }
+
+  private createUserScheduleIfNotCreated(userId) {
+    if (this.doesUserHaveASchedule(userId)) {
+      return;
+    }
+
+    this.schedule.set(userId, new UserSchedule({ timeLimit: this.timeLimit, userId }));
+  }
+
+  private getUserSchedule(userId) {
+    return this.schedule.get(userId);
+  }
+
+  public async assignWorkspaceToUser(userId, workspace, startAtTimestamp = Date.now()) {
+    if (!this.doesUserHaveASchedule(userId)) {
+      this.createUserScheduleIfNotCreated(userId);
+    }
+
+    const userSchedule = this.getUserSchedule(userId);
+    userSchedule.assignWorkspace(workspace, startAtTimestamp);
+    const rootParent = await this.rootParentCache.getRootParentOfWorkspace(workspace);
+    this.cacheForWhenTreeLastWorkedOn[rootParent.id] = startAtTimestamp;
+  }
+
+  public getMostRecentAssignmentForUser(userId) {
+    if (!this.doesUserHaveASchedule(userId)) {
+      return undefined;
+    }
+
+    const userSchedule = this.getUserSchedule(userId);
+    return userSchedule.getMostRecentAssignment();
+  }
+
+  /*
+    NOTE: this only considers the trees associated with the rootWorkspaces
+    argument
+  */
+  public async isInTreeWorkedOnLeastRecently(rootWorkspaces, workspace) {
+    const treesWorksOnLeastRecently = this.getTreesWorkedOnLeastRecently(rootWorkspaces);
+    const rootParent = await this.rootParentCache.getRootParentOfWorkspace(workspace);
+    return !!(treesWorksOnLeastRecently.find(rootWorkspace => rootWorkspace === rootParent));
+  }
+
+  /*
+    NOTE: this only considers the trees associated with the rootWorkspaces
+    argument
+  */
+  public getTreesWorkedOnLeastRecently(rootWorkspaces) {
+    const treesNotYetWorkedOn = rootWorkspaces.filter(
+      rootWorkspace => this.cacheForWhenTreeLastWorkedOn[rootWorkspace.id] === undefined
```

Comment:
In cases like this, I prefer if the filter uses a var name like ``r`` instead of ``rootWorspace``. We can tell that it is a rootWorkspace from reading 9 characters back.